faster and parallel version of sampleEntropy

mbstadler opened the pull request:
This is an updated version of `sampleEntropy` with some performance optimizations:

- one-pass calculation of mean/sd using a single for loop (no need for both `std::accumulate` and `std::for_each`)
- use pointer arithmetic to reduce repeated index arithmetic (e.g. `xi = &data[i]` and `xi[k]` instead of `data[i+k]`)
- branch pruning (`if(eq) ...`)
- parallelize over outermost for loop (i) using OpenMP (iterations are mostly independent, just need to be careful when incrementing `Cm` and `Cm1` -> use `reduction`)

The new version of `sampleEntropy` can be used as a drop-in replacement of the old one and gives identical results.

Currently, the parallelization is not made use of in `footprintR` (need to check if this would not lead to nested parallelization).

mbstadler committed:
faster and parallel version of sampleEntropy

mbstadler commented:
In my benchmarks (on an M1 MAX arm64 CPU), the new version is 15-30% faster with `nThreads = 1` and achieves near linear speed-ups up to 8 threads (7.6-fold faster compared to single thread).

csoneson commented:
This looks good to me!

ppapasaikas committed:
sampleEntropy2:

Even faster sample entropy calculation.
The original function compares every pair (i,j) of length-m segments.
By presorting the values and hashing their indices we can limit the i,j queries to only those js whose first value is within err.

This gives an ~6x speedup

ppapasaikas commented:
Pushed a sampleEntropy2 function that has all of Michael's improvements but also a change in the logic of the code that speeds it up another 6x (on top). 

Briefly:.
The original function compares every pair of length-m segments and counts how many pairs are within  err at all m positions. 
Now, we use a pre-filter: if two segments match at all m positions, they must at least match at the first position. We sort all segment starts by their first value and, for each start i, we only consider those starts j whose first value is within ±err. Everything else is same as before. This skips a huge number of pairs, produces identical results, and is much faster.

Kept if for now as a separate function because the code logic is different. We can replace if you agree.

ppapasaikas commented:
I think there is also significant gains to be made (if we want) by a simple sampling approach. Currently all i possible starting positions are considered for the i,j comparisons. This is ok for sequences <5K but blows up fast. A single read of length 100K will bring the function to a crawl. We can cap the possible number of starting positions i to some number maxI (can be a parameter with a default around 1K).

This of course will not produce identical results but in practice it will not make a difference and can produce another big speed up. 
What do you think?

mbstadler committed:
fuse sampleEntropy2 into sampleEntropy

mbstadler commented:
Good morning @ppapasaikas! That looks great! I have run my benchmarking set and for not too long data vectors (~1,000 elements) it's a great speedup (~30-50% on my test data and hardware). It runs a bit slower on longer vectors (about 20%, ~10,000 elements), likely because the overhead to allocate/sort the pairs list start to take more time, but it seems that this will not be typical for our applications if our reads don't get much longer.

I think adding a parameter to cap the number of observed pairs seems a great idea, because I have seen it run for hours on an individual read before. If we set the default value to `Inf` and call it within `footprintR` with something reasonable (maybe 10,000?) that should make it run robustly.

I am pushing a few formatting suggestions, and also move/rename `sampleEntropy2` to `sampleEntropy` (no need to have two versions in parallel).

Feel free to add the cap argument if you like!

ppapasaikas committed:
Expose SampleEntropy control arguments

ppapasaikas committed:
Default maxStarts 1000 for Entropy calculation

ppapasaikas committed:
added maxStarts parameter to control number of starting positions sampled during SampleEntropy calculation.

Sampling uses a fixed seed so results should be reproducible

ppapasaikas committed:
Added the sampleEntropy maxStarts -1 default

ppapasaikas committed:
Adapt the maxEntropy threshold so that the filterReadsBam maxStarts 1000 default still filters 1 read.

ppapasaikas committed:
Generate sampleEntropy documentation
Add to calcReadStats documentation for EntrControl list argument

ppapasaikas committed:
Do not run examples for sampleEntropy (need the documentation as it is referenced in calcReadStats but want to keep it internal)

ppapasaikas committed:
Switch from sampling to even-spaced maxStarts. Ensures consistent results across platforms.

ppapasaikas commented:
Pushed code  implementing the cap argument (maxStarts). When cap is less than the n of measurements  then evenly spaced starting positions are picked. The random sampling policy was giving inconsistent results across platforms.

 The sampleEntropy control arguments (m,r,maxStarts,nThreads) are now also exposed to calcReadStats with an added argument (EntrControl) - a named list-. 

Function default and calcReads default set to -1 (no cap).
Default maxStarts for filterReadsBam set to 1000.

Some slight modification in the tests to reflect the above changes.

Below a comparison of Entr values without cap and with a cap at 1K on  ~1500 reads and also of their Delta as a function of the number of measured positions on the input reads. The reduction in runtime is significant in the same test (from 30" to 13")
<img width="1500" height="800" alt="maxStarts_comp" src="https://github.com/user-attachments/assets/c3e5873a-8bae-42cb-9319-bcc0085884b5" />

mbstadler commented:
Thanks @ppapasaikas for all these additions!

I am not sure it's good to have different default parameters (`maxStarts`) for different functions (e.g. `calcReadStats` and `filterReadsBam`). Also, it is not clear if it would be possible to even change them, as they may not always be exposed.

From your above analysis it seems that maybe we can make `maxStarts=1000` the default everywhere, what do you think?

Regarding `sampleEntropy`: I would prefer not to use any `\dontrun{}` in the examples. Would it be an option to make `sampleEntropy` an exported, public function? That would also be better given that it is referred to in `calcReadStats`, and the function may be useful in other contexts outside of modified bases?

ppapasaikas commented:
I agree that maxStarts 1000 looks like a reasonable choice globally especially since the larger Deltas appear to concentrate on the mid-low register of entropy values (which makes sense, those would be reads with a a lot of structured regions so missing some of those can have an impact), and the metric is used for filtering out the high entropy ones.  Will do that.


Re. exporting sampleEntropy. Of course this is an option. My thinking was that since this is a function that is not footprinting specific + does not accept input in SExp format it might be better to keep internal. 
Related to this: I do not understand why when the function is internal and called with `:::` in the example it's not recognized (which was the whole reason of wrapping around ` \dontrun{} `

ppapasaikas committed:
Set maxStarts to 1000 globally.
Export sampleEntropy

ppapasaikas committed:
Export sampleEntropy

ppapasaikas committed:
remove dontrun from sampleEntropy example

ppapasaikas committed:
Namespace includes sampleEntropy

mbstadler committed:
remove unnecessary namespace specifier

csoneson committed:
Fix merge conflicts

Merge branch 'devel' into faster-sample-entropy

# Conflicts:
#	R/calcReadStats.R
#	R/filterReadsBam.R
#	man/filterReadsBam.Rd
#	tests/testthat/test-calcReadStats.R
#	tests/testthat/test-filterReadsBam.R

csoneson commented on R/calcReadStats.R:
```diff
@@ -472,6 +479,23 @@ calcReadStats <- function(se,
     .assertVector(x = BPPARAM, type = "BiocParallelParam")
     .assertScalar(x = verbose, type = "logical")
 
+    if (is.null(EntrControl)) {
```
Need to check here that `EntrControl` is a list (or `NULL`) - will get an error later if it's a vector for example.

csoneson replied:
Added this check in https://github.com/fmicompbio/footprintR/pull/56

csoneson commented on R/calcReadStats.R:
```diff
@@ -472,6 +479,23 @@ calcReadStats <- function(se,
+        EntrControl <- list()
+    }
+
+    SEctrl <- modifyList(
```
Do we need a new object `SEctrl` here, or could we just overwrite `EntrControl`?

csoneson replied:
Fixed in 4afa08e

csoneson commented on src/filter_modbam_cpp.cpp:
```diff
@@ -344,7 +344,7 @@ Rcpp::NumericVector filter_modbam_cpp(std::string infile,
         if (maxEntropy >= 0) {
             // calculate sample entropy
-            if (mod_probs.size() == 0 || sampleEntropy(mod_probs, 2, 0.2) > maxEntropy) {
+            if (mod_probs.size() == 0 || sampleEntropy(mod_probs, 2, 0.2, 1000) > maxEntropy) { 
```
explicitly provide nThreads?

csoneson replied:
Added in 32f8674

csoneson commented on tests/testthat/test-sampleEntropy.R:
```diff
@@ -22,9 +22,9 @@ test_that("sampleEntropy works", {
+    expect_equal(sampleEntropy(x, 2, 0.5, -1), 1.27978397948638)
+    expect_equal(sampleEntropy(y, 3, 0.2, -1), 0.496944555201614)
+    expect_equal(sampleEntropy(z, 2, 0.2, -1), 0.014575909171002, tolerance = 1e-2)
+    expect_identical(sampleEntropy(allNA, 2, 0.2, -1), 0)
```
Add tests with maxStarts != -1?

csoneson replied:
Added in e1352fc

csoneson commented on src/sampleEntropy.cpp:
```diff
@@ -17,59 +22,144 @@ using namespace Rcpp;
+    // Number of possible Starting positions.
+    // Each subsequence is of length m, so the last possible start is at index N - m.
+    const unsigned int S = N - m;
+    if (S <= 1) return 0.0;
```
This should never happen I think - same check as on line 54

csoneson replied:
Removed in 58d6ba2

csoneson commented on src/sampleEntropy.cpp:
```diff
@@ -17,59 +22,144 @@ using namespace Rcpp;
+    if (!std::isfinite(maxStarts) || maxStarts <= 0) {
+        maxI = S;  // use all starts
+    } else {
+        maxI = std::min(S, static_cast<unsigned int>(maxStarts));
```
I think we don't cast things this explicity in other places - should we be consistent?

csoneson replied:
Replaced in 18e2c3d

csoneson commented on src/sampleEntropy.cpp:
```diff
@@ -17,59 +22,144 @@ using namespace Rcpp;
+    if (maxI < S) {
+        std::vector<unsigned int> subset;
+        subset.reserve(maxI);
+        for (size_t t = 0; t < static_cast<size_t>(maxI); ++t) {
```
For consistency (here and in a few other places further down): ++t -> t++?

csoneson replied:
Changed in 0a39485

csoneson commented on src/sampleEntropy.cpp:
```diff
@@ -17,59 +22,144 @@ using namespace Rcpp;
+            size_t idx = (t * static_cast<size_t>(S)) / static_cast<size_t>(maxI);
+            subset.push_back(static_cast<unsigned int>(idx));
+        }
+        starts.swap(subset);
```
Do we need the swap, or just a redefinition of starts?

csoneson replied:
TBD

csoneson replied:
Simplified in d75d809

csoneson reviewed (commented):
Just went through this PR again - looks good to me, just added a few comments.

csoneson committed:
Replace macos-13 GHA runner (will be retired in Dec 2025)

csoneson committed:
Add 'regions' argument to filterPositions

csoneson committed:
Check that EntrControl is a list

csoneson committed:
Redefine EntrControl instead of creating new object SEctrl

csoneson committed:
Explicitly provide nThreads to sampleEntropy

csoneson committed:
Add unit tests for sampleEntropy with maxStarts != -1

csoneson committed:
Remove unneeded check

csoneson committed:
Replace static_cast for consistency

csoneson committed:
Use i++ instead of ++i in for loops for consistency

csoneson committed:
Merge pull request #56 from fmicompbio/add-region-arg-filterpos

Add 'regions' argument to filterPositions

csoneson committed:
Simplify generation of start positions

Co-authored-by: Michael Stadler <stadler.michael@gmail.com>
Co-authored-by: Charlotte Soneson <charlottesoneson@gmail.com>

csoneson committed:
Bump version, update NEWS

Co-authored-by: Michael Stadler <stadler.michael@gmail.com>
Co-authored-by: Charlotte Soneson <charlottesoneson@gmail.com>
